Full review

YawKar opened the pull request.

YawKar committed:
- 1st (0-based) commit

YawKar committed:
- Add CharactersController and ComicsController
- Add dummy @GetMappings for both of them

YawKar committed:
- Add Swagger support
- Add SpringDoc OpenAPI Web Utility

YawKar committed:
- Change Java language version
- Add README.MD

YawKar committed:
- Change SQL dialect to PostgreSQL95Dialect (because Hibernate's version is 5.6.9.Final)
- Add Character entity
- Add Comic entity
- Add many-to-many connection between them
- Add CRUD CharactersRepository
- Add CRUD ComicsRepository

YawKar committed:
- Rename Character to Hero (dumb mistake)
- Rename CharactersRepository to HeroesRepository
- Rename charactersPresent to heroesPresent in Comic entity
- Rename accessors in Comic
- Rename table's and columns' names in Hero entity

YawKar committed:
- Removed unnecessary cascading in Comics

YawKar committed:
- Remove `synopsis` from Comic
- Add ComicDTO with constraints on fields
- Add HeroDTO with constraints on fields
- Add spring-boot-starter-validation to pom.xml

YawKar committed:
- Renamed Character to Hero in CharactersController

YawKar committed:
- Add messages for validation constraints
- Add JsonProperty annotation to `realName`, `executiveEditor` and `coverArtists` fields

YawKar committed:
- Replace occurrences of 'Character' to 'Hero' in ComicsController
- Add postNewComic handler to ComicsController

YawKar committed:
- Add HeroesService
- Add ComicsService
- Add ModelMapper dependency to pom.xml

YawKar committed:
- Add ModelMapperConfiguration
- Add Bean of ModelMapper to ModelMapperConfiguration
- Inject ModelMapper into services

YawKar committed:
- Add `insertNewHero` method to `HeroesService`
- Add `insertNewComic` method to `ComicsService`
- Update `postNewHero` method in `CharactersController`
- Update `postNewComic` method in `ComicsController`
- Inject `HeroesService` to `CharactersController`
- Inject `ComicsService` to `ComicsController`

YawKar committed:
- Add `id` field to HeroDTO and ComicDTO

YawKar committed:
- POST requests return DTOs with id

YawKar committed:
- Implement all GET handlers in CharactersController
- Add appropriate methods to HeroesService

YawKar committed:
- Implement all GET handlers in ComicsController
- Add appropriate methods to ComicsService

YawKar committed:
- Replace `List` with `Set` in many-to-many fields of entity classes
- Add `equals()` and `hashCode()` to entity classes

YawKar committed:
- Add `PUT putHero()` handler method that updates a hero from dto
- Add `PUT putComic()` handler method that updates a comic from dto
- Add `entity.setId(0L)` to insert methods of services in order to force database to create new entities
- Add `updateComic()` to ComicsService
- Add `updateHero()` to HeroesService

YawKar committed:
- Add `AttachmentDTO` for connecting heroes with comics

YawKar committed:
- Bugfix: Remove using of nested fields in `equals()` and `hashCode()` methods of entities that produced StackOverflowException in Hibernate
- Add `AttachmentsController`
- Add `PUT attachComicsToHeroes()` handler to `AttachmentsController`
- Add `AttachmentsService`
- Add `attachComicsToHeroes()` to `AttachmentsService`

YawKar committed:
- Add `SwaggerConfig` with OpenAPI definition bean
- Add `Schema` descriptions in DTOs

YawKar committed:
- Add Swagger documentation annotations to `AttachmentsController`
- Add Swagger documentation annotations to `CharactersController`
- Add Swagger documentation annotations to `ComicsController`

YawKar committed:
- Update README.MD

YawKar committed:
Update README.md

- Update README.md
- Remove extra horizontal lines

YawKar committed:
Merge branch 'master' into full_review

46551118 commented on src/main/java/com/yawkar/marvelcatalog/service/ComicsService.java:
```diff
@@ -0,0 +1,58 @@
+        this.modelMapper = modelMapper;
+    }
+
+    public ComicDTO insertNewComic(ComicDTO comicDTO) {
+        Comic comic = modelMapper.map(comicDTO, Comic.class);
+        comic.setId(0L);
+        return modelMapper.map(comicsRepository.save(comic), ComicDTO.class);
+    }
+
+    public ComicDTO updateComic(ComicDTO comicDTO) {
+        Comic comic = comicsRepository.findById(comicDTO.getId())
+                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND));
+        modelMapper.map(comicDTO, comic);
+        return modelMapper.map(comicsRepository.save(comic), ComicDTO.class);
+    }
+
+    public List<ComicDTO> findAllComics() {
+        List<ComicDTO> comicDTOS = new ArrayList<>();
+        comicsRepository.findAll().forEach(comic -> comicDTOS.add(modelMapper.map(comic, ComicDTO.class)));
+        return comicDTOS;
+    }
+
+    public ComicDTO findComicById(long comicId) {
+        return modelMapper.map(
+                comicsRepository.findById(comicId).orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND)),
+                ComicDTO.class);
+    }
+
+    public List<HeroDTO> findAllHeroesByComicId(long comicId) {
+        List<HeroDTO> heroDTOS = new ArrayList<>();
+        comicsRepository.findById(comicId).orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND))
+                .getHeroesPresent()
+                .forEach(hero -> heroDTOS.add(modelMapper.map(hero, HeroDTO.class)));
+        return heroDTOS;
```
Уровень бизнес-логики и представления очень связан

Сервис должен работать только с сущностью и не должен беспокоится о том, чтобы перегонять сущность из одного вида в другую

Это задача контроллера - он должен отправлять на вход в сервис адаптер который реализует тот же интерфейс, что реализует сущность, а затем ответ сервиса маппить на необходимое представление

46551118 commented on src/main/java/com/yawkar/marvelcatalog/service/AttachmentsService.java:
```diff
@@ -0,0 +1,51 @@
+    }
+
+    @Transactional
+    public void attachComicsToHeroes(AttachmentDTO attachmentDTO) {
+        for (long heroId : attachmentDTO.getHeroIds()) {
+            if (!heroesRepository.existsById(heroId)) {
+                throw makeNotFound("Hero with hero_id = %d not found", heroId).get();
+            }
+        }
+        for (long comicId : attachmentDTO.getComicIds()) {
+            if (!comicsRepository.existsById(comicId)) {
+                throw makeNotFound("Comic with comic_id = %d not found", comicId).get();
+            }
+        }
+        attachmentDTO.getHeroIds().forEach(heroId -> {
+            Hero hero = heroesRepository.findById(heroId).orElseThrow(makeNotFound("Hero with hero_id = %d not found", heroId));
+            attachmentDTO.getComicIds().forEach(comicId ->
+                hero.getComicsInWhichPresent().add(
+                        comicsRepository.findById(comicId).orElseThrow(makeNotFound("Comic with comic_id = %d not found", comicId))
+                )
+            );
+            heroesRepository.save(hero);
+        });
```
это не должно быть в отдельном сервисе, а в сервисе для героев, реализовано слишком сложно, можно сделать проще

сервис удалить, логику переписать, подумай сначала над связями между объектами, а только затем пиши код

46551118 commented on src/main/java/com/yawkar/marvelcatalog/service/HeroesService.java:
```diff
@@ -0,0 +1,58 @@
+        this.modelMapper = modelMapper;
+    }
+
+    public HeroDTO insertNewHero(HeroDTO heroDTO) {
+        Hero hero = modelMapper.map(heroDTO, Hero.class);
+        hero.setId(0L);
+        return modelMapper.map(heroesRepository.save(hero), HeroDTO.class);
+    }
+
+    public HeroDTO updateHero(HeroDTO heroDTO) {
+        Hero hero = heroesRepository.findById(heroDTO.getId())
+                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND));
+        modelMapper.map(heroDTO, hero);
+        return modelMapper.map(heroesRepository.save(hero), HeroDTO.class);
+    }
+
+    public List<HeroDTO> findAllHeroes() {
+        List<HeroDTO> heroDTOS = new ArrayList<>();
+        heroesRepository.findAll().forEach(hero -> heroDTOS.add(modelMapper.map(hero, HeroDTO.class)));
+        return heroDTOS;
+    }
+
+    public HeroDTO findHeroById(long heroId) {
+        return modelMapper.map(
+                heroesRepository.findById(heroId).orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND)),
+            HeroDTO.class);
+    }
+
+    public List<ComicDTO> findAllComicsByHeroId(long heroId) {
+        List<ComicDTO> comicDTOS = new ArrayList<>();
+        heroesRepository.findById(heroId).orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND))
+                .getComicsInWhichPresent()
+                .forEach(comic -> comicDTOS.add(modelMapper.map(comic, ComicDTO.class)));
+        return comicDTOS;
```
тоже самое

46551118 commented on src/main/java/com/yawkar/marvelcatalog/repository/HeroesRepository.java:
```diff
@@ -0,0 +1,7 @@
+import com.yawkar.marvelcatalog.entity.Hero;
+import org.springframework.data.repository.CrudRepository;
+
+public interface HeroesRepository extends CrudRepository<Hero, Long> {
```
используй JpaRepository, там больше полезных методов уже реализовано

46551118 commented on src/main/java/com/yawkar/marvelcatalog/repository/ComicsRepository.java:
```diff
@@ -0,0 +1,7 @@
+import com.yawkar.marvelcatalog.entity.Comic;
+import org.springframework.data.repository.CrudRepository;
+
+public interface ComicsRepository extends CrudRepository<Comic, Long> {
```
используй JpaRepository, там больше полезных методов уже реализовано

46551118 commented on src/main/java/com/yawkar/marvelcatalog/entity/Hero.java:
```diff
@@ -0,0 +1,83 @@
+
+    public Hero() {}
+
+    public long getId() {
+        return id;
+    }
+
+    public void setId(long id) {
+        this.id = id;
+    }
+
+    public String getRealName() {
+        return realName;
+    }
+
+    public void setRealName(String realName) {
+        this.realName = realName;
+    }
+
+    public String getAlias() {
+        return alias;
+    }
+
+    public void setAlias(String alias) {
+        this.alias = alias;
+    }
+
+    public List<String> getSuperpowers() {
+        return superpowers;
+    }
+
+    public void setSuperpowers(List<String> superpowers) {
+        this.superpowers = superpowers;
+    }
+
+    public Set<Comic> getComicsInWhichPresent() {
```
для этого есть ломбок

46551118 commented on src/main/java/com/yawkar/marvelcatalog/entity/Hero.java:
```diff
@@ -0,0 +1,83 @@
+    @Column(name = "real_name")
+    private String realName;
+    private String alias;
+    @ElementCollection
+    private List<String> superpowers;
```
плохая практика

либо создай объект суперсилы многие-ко-многим, либо сделай это одной строкой с тегами

46551118 commented on src/main/java/com/yawkar/marvelcatalog/entity/Comic.java:
```diff
@@ -0,0 +1,79 @@
+
+@Entity
+@Table(name = "comics")
+public class Comic {
```
ломбок, элемент коллекшн, то же самое

46551118 commented on src/main/java/com/yawkar/marvelcatalog/dto/HeroDTO.java:
```diff
@@ -0,0 +1,59 @@
+import java.util.List;
+
+@Schema(description = "A hero entity")
+public class HeroDTO {
```
ломбок

46551118 commented on src/main/java/com/yawkar/marvelcatalog/dto/ComicDTO.java:
```diff
@@ -0,0 +1,62 @@
+import java.util.List;
+
+@Schema(description = "A comic entity")
+public class ComicDTO {
```
ломбок

46551118 commented on src/main/java/com/yawkar/marvelcatalog/dto/AttachmentDTO.java:
```diff
@@ -0,0 +1,41 @@
+import java.util.List;
+
+@Schema(description = "An attachment entity")
+public class AttachmentDTO {
```
ломбок

46551118 commented on src/main/java/com/yawkar/marvelcatalog/controller/ComicsController.java:
```diff
@@ -0,0 +1,161 @@
+        return comicsService.findAllHeroesByComicId(comicId);
+    }
+
+    @Operation(
+            summary = "Posts new comic",
+            description = "Posts new comic and returns the same object with `id` property that represents the internal id" +
+                    " of the comic"
+    )
+    @ApiResponses(
+            {
+                    @ApiResponse(
+                            responseCode = "200",
+                            description = "Successfully posted new comic",
+                            content = @Content(
+                                    mediaType = MediaType.APPLICATION_JSON_VALUE,
+                                    schema = @Schema(implementation = ComicDTO.class)
+                            )
+                    )
+            }
+    )
+    @PostMapping
+    public ComicDTO postNewComic(@Valid @RequestBody ComicDTO comicDTO) {
```
читать это - пиздец

пиши по-человечески

46551118 commented on src/main/java/com/yawkar/marvelcatalog/controller/AttachmentsController.java:
```diff
@@ -0,0 +1,50 @@
+            }
+    )
+    @PutMapping
+    public ResponseEntity<Void> attachComicsToHeroes(@Valid @RequestBody AttachmentDTO attachmentDTO) {
```
перенести в контроллер героев

46551118 commented on src/main/java/com/yawkar/marvelcatalog/configuration/ModelMapperConfiguration.java:
```diff
@@ -0,0 +1,14 @@
+
+    @Bean
+    public ModelMapper getModelMapper() {
+        return new ModelMapper();
```
а вот мапперы лучше руками писать

46551118 reviewed (changes requested):
Исправить:

- Связь между уровнями (почитай про представление - сервис - репозиторий чтобы понять как правильно это делать)
- Написать тесты на сервисы, а лучше сразу на уровень апи
- Написать тесты на мапперы
- Убрать объект маппер и написать мапперы руками

YawKar committed:
- Remove Attachments-related controller, dto and service (will be re-implemented in future)
- Change CrudRepository to JpaRepository in ComicsRepository and HeroesRepository because it provides more methods
- Add ComicsService interface
- Add HeroesService interface
- Clear methods' bodies in HeroesServiceImpl and ComicsServiceImpl to re-implement them in future
- Add ComicView and HeroView

YawKar committed:
Merge pull request #2 from YawKar/master

Update README.md

YawKar committed:
- Add Lombok
- Remove modelmapper:3.1.0
- Add my own ModelMapper
- Inject ModelMapper into controllers
- Refactor DTOs and entities with Lombok annotations

YawKar committed:
- Change `superpowers` and `coverArtists` fields' type from List<String> to String to store them in the same table
- Populate ComicView and HeroView with fields and Lombok annotations

YawKar committed:
- Add implementations to ModelMapper methods

YawKar committed:
- Change `updateHero()` signature: add Long heroId to support RESTful PUT method convention in CharactersController
- Implement all methods in CharactersController
- Change route of `putHero()` to `characters/{heroId}` because that way it acts like a collection

YawKar committed:
- Change `updateComic()` signature: add Long heroId to support RESTful PUT method convention in ComicsController

- Implement all methods in ComicsController
- Change route of `putComic()` to `comics/{comicId}` because that way it acts like a collection

YawKar committed:
- Add PUT method for updating the list of comics in which the specified hero presents
- Remove generic <T> from HeroesService and ComicsService interfaces (now T's always long)

YawKar committed:
- Add `HeroNotFoundException`
- Implement all methods of `HeroesServiceImpl`

YawKar committed:
- Add `ComicNotFoundException`
- Implement all methods of `ComicsServiceImpl`
- Remove `id` field from DTOs and therefore change setters/getters calls in ModelMapper and Services

YawKar committed:
- Fix mistake `hero`->`heroDTO` in ModelMapper

YawKar committed:
- Add ExceptionView for uniform view of custom exceptions
- Add MarvelCatalogExceptionHandler for handling custom exceptions

YawKar committed:
chore: restructure packages

YawKar committed:
refactor: delete ModelMapper and create abstract GeneralMapper with shared methods

Add ComicMapper and HeroMapper. Add AllArgsConstructor to controllers.

YawKar committed:
feat: add chained lombok accessors to all data classes

YawKar committed:
test: add ComicMapperTest with tests of all methods

YawKar committed:
test: remove unnecessary setups

YawKar committed:
test: add HeroMapperTest with all methods tests

YawKar committed:
test: add ComicsServiceImplTest with one test

YawKar committed:
test: add HeroesServiceImplTest with one test

also change BDDMockito to Mockito in ComicsServiceImplTest

YawKar committed:
test: add test classes for ComicsController and CharactersController
